src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+        internal sealed class SnapshotManager
+        {
+            /// <summary>
+            /// The int key corresponds to <see cref="Snapshot.GlobalStateIndex"/>.
+            /// </summary>
+            private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
+            private readonly ImmutableDictionary<int, Snapshot> _incrementalSnapshots;
+
+            private SnapshotManager(ImmutableArray<SharedWalkerState> walkerGlobalStates, ImmutableDictionary<int, Snapshot> incrementalSnapshots)
+            {
+                _walkerGlobalStates = walkerGlobalStates;
+                _incrementalSnapshots = incrementalSnapshots;
+            }
+
+            internal NullableWalker RestoreWalkerToAnalyzeNewNode(
+                int position,
+                BoundNode nodeToAnalyze,
+                CSharpCompilation compilation,
```

Comment:
>CSharpCompilation compilation, [](start = 16, length = 30)

Do we need to pass this in? Is the binder always available? I think we can get compilation from the binder.